### ISSUE TITLE
Launch Clear Sky

### ISSUE THREAD

**Issue by Xottab-DUTY** [OpenXRay/xray-16#382]
Call of Pripyat engine which we are using wasn't changed too much since Clear Sky.

We can launch it! Lets do this!

Status:
- [x] Alpha (many bugs, crashes, you can't finish the game)
- [x] Beta (bugs, crashes, but you can finish the game)
- [x] Release Candidate (there are minor bugs, from a few to none crashes, game is stable finishable)
- [ ] Release (game is stable)

Tasks:
- [x] Renderer
  - [x] Don't crash on missing COP functionality. (c7dec853de8899e1ad98a0758e2dd596cea265fa)
  - [x] R2/R3/R4 issues with old/new shadow cascades (d6780cc2bc0c0fe385d78c151118f49bd207cff9)
  - [x] R3 issue with mark_msaa_edges.ps (8e078286ceca956713ae59db50a79a277f4a093f)
  - [x] Launch R4. (requires COP shaders to work correctly, otherwise expect bugs)
- [x] Lua Scripting
  - [x] Return deleted script export
- [x] Main menu
  - [x] Exit
  - [x] Titles
  - [x] Options
  - [x] Load game
- [x] UI
  - [x] Actor inventory UI
  - [x] PDA
    - [x] Weird background stripes due to `ui_pda2_fr2` texture fragments being in different files.
          It was supported in CS, but `CUIFrameLineWnd` was simplified in COP and this feature was removed. (885e2c1c2f123539bce74ac17836579dd59c6fcf)
    - [x] Nimble's PDA usage tutorial (c36d59e49de080d95c1414c66e5af736240bd188, c7693a65452bbef3840bfd8f18e826f8989f38cf)
  - [x] Elite detector doesn't show artefacts (e1a4a46042395966e361a340a595b1f978354c0d)
  - [x] Slightly different upgrade system (e2ef829a61c0b8dedb393fc87693a81cfbb638fa, 2f9bbe5985807b1426fda11d8f944ac38c44a52c)
  - [ ] Return deleted UI functionality and script export
- [x] Gameplay
  - [x] Monsters inventory (670632a6126f411fa25385967c5f0590e52fcd21, 3dc9f5a846bf637b1f6091df60189c2120864a29)
  - [x] Clear Sky task system
  - [x] Fix some quest NPCs can't enter their animation smart covers (e.g. Cold, Novikov on Clear Sky base) (52ce00154a7675f517cccfbeeac7ff903094bc48)
  - [ ] Few missing features from weapons and probably something else
  - [ ] Mutants AI
- [x] Singleplayer
  - [x] Launch it
  - [x] Finish the game
  - [ ] Launch and finish some global mods
    - [ ] SRP
    - [ ] SRP-Lite
    - [ ] OGSM
- [x] Multiplayer (no full support)
  - [x] Traffic compression unavailable (9f8b4e95ff8dbb9d53ae92df8169fcb9022d54c6)
  - [x] Multiplayer-specific UI works except missing captions from `ui_custom_msgs.xml` (bb894ade631a1242abf4c9cb554c8f00e645e277).
    - You have to import `ui_custom_msgs.xml` from COP yourself.

**Comment by collectioner** [comment]
@Xottab-DUTY please, create a small instruction how we can launch it. Simply copy binaries from OXR and replace original? Need guide.

**Comment by Xottab-DUTY** [comment]
Launch it as usual with this gamedata:
[gamedata.zip](https://github.com/OpenXRay/xray-16/files/3145212/gamedata.zip)

**Comment by Xottab-DUTY** [comment]
Main menu was launched quite a few months ago, but if someone interested:

![ss_Султан_05-05-19_12-17-13_(mainmenu)](https://user-images.githubusercontent.com/2989212/57190071-c3c31980-6f2f-11e9-9ea7-78fe3fd69992.jpg)
(You can see red text on top of the screenshot, that's because screenshot was made on a Mixed build)

**Comment by Xottab-DUTY** [comment]
Just finished the game two days ago.
![image](https://user-images.githubusercontent.com/2989212/59978337-c0294600-95f4-11e9-83e1-a209860c0194.png)

**Comment by Xottab-DUTY** [comment]
Okay, it's almost done. There's still small problems remain, but you can play it.

**Comment by CasualDev242** [comment]
Can you please list the differences between the normal gamedata from this repository and the CS one? I ask because when comparing with WinMerge, although there are multiple differences, it can easily be due to some files being outdated. As far as I can tell, the important thing is having the game changed to "cs" in openxray.ltx right?

**Comment by johndoe71rus** [comment]
need create logo to start (and for SoC)

**Comment by Xottab-DUTY** [comment]
New gamedata for Clear Sky. With it you can launch R2/R3/R4 freely.
[gamedata.zip](https://github.com/OpenXRay/xray-16/files/4313569/gamedata.zip)

**Comment by Xottab-DUTY** [comment]
New gamedata. Removed shaders with new COP shadow cascades, now engine will use old SOC/CS ones.
[gamedata.zip](https://github.com/OpenXRay/xray-16/files/4348865/gamedata.zip)

**Comment by Xottab-DUTY** [comment]
Updated gamedata for [558 release](https://github.com/OpenXRay/xray-16/releases/tag/558).
[gamedata.zip](https://github.com/OpenXRay/xray-16/files/4502665/gamedata.zip)

**Comment by phantasm** [comment]
can't start a new game or open settings menu. using GOG version of CS on Win10 x64, GTX 1070 with latest driver (445.87)
[OpenXRay_error_report_200422-040425.zip](https://github.com/OpenXRay/xray-16/files/4513252/OpenXRay_error_report_200422-040425.zip)

**Comment by globekat** [comment]
After resolving the previous issues, I've jumped back in and logged some more. Mainly to do with the Rendering modes.
- DX8 - no issues.
- DX9 (all modes) - crash when loading/creating a game. Attached screenshots showing error dialogs. The same DX9 render options work fine in CoP.
- DX10, DX11 - additional shader related error when starting Clear Sky. Also everything is very dark. See attached screenshots (screenshot is DX10, but DX11 looks the same).
- OpenGL - see screenshot. Not sure what's going on here, but its definitely shader related.

Computer/System Information
- OS: Windows 10 1909
- OpenXRay build version [December 2020 Preview, Mixed]
- Clear Sky: Steam, 1.5.10
- Graphics Card: Nvidia GeForce GTX 970, Version 456.71

[openxray_alex_r2.log](https://github.com/OpenXRay/xray-16/files/6135898/openxray_alex_r2.log)
[openxray_alex_r2a.log](https://github.com/OpenXRay/xray-16/files/6135899/openxray_alex_r2a.log)
[openxray_alex_r3.log](https://github.com/OpenXRay/xray-16/files/6135900/openxray_alex_r3.log)
[openxray_alex_r4.log](https://github.com/OpenXRay/xray-16/files/6135901/openxray_alex_r4.log)
[openxray_alex_r25.log](https://github.com/OpenXRay/xray-16/files/6135902/openxray_alex_r25.log)
[openxray_alex_rgl.log](https://github.com/OpenXRay/xray-16/files/6135903/openxray_alex_rgl.log)
DX9 crash dialogs when loading a game.
![Screenshot 2021-03-14 14 27 39](https://user-images.githubusercontent.com/13399222/111054518-327d4080-84d2-11eb-8a63-68e4ba115ae0.png)
![Screenshot 2021-03-14 14 27 41](https://user-images.githubusercontent.com/13399222/111054521-4032c600-84d2-11eb-929c-adbb112c5453.png)
DX10/11 very deep shadows. Also some weird shading on the ceiling (cloud textures seem to be leaking through?).
![Screenshot 2021-03-14 14 04 34](https://user-images.githubusercontent.com/13399222/111054529-4d4fb500-84d2-11eb-8ded-9ae3d6ca0abd.png)
OpenGL shader glitches?
![20210314143048_1](https://user-images.githubusercontent.com/13399222/111054547-7a9c6300-84d2-11eb-8de8-0781dc5784f5.jpg)

**Comment by Xottab-DUTY** [comment]
@Setakat, egh, I thought I've fully fixed that lighting issues... Could you share your user.ltx here?

Will try to fix it)

**Comment by globekat** [comment]
Attached user.ltx as requested - [user.ltx.txt](https://github.com/OpenXRay/xray-16/files/6153947/user.ltx.txt)

Only 1 crash so far - occurred when I decided to try and drag a primary weapon into the secondary weapon slot, since CoP has flexible slots compared to SoC/CS. Complained about insufficient grid space.
Looking at it, I can see two possible resolutions
- Add additional logic engine side regarding what weapons can be equipped in each slot so flexible slots are only for CoP. SoC/CS will restrict primary/secondary weapons to their respective slots only
- Mod the gamedata to increase the size of the slot to 2x6 in the background to prevent a crash, and provide an optional inventory redesign that makes it apparent that two primary sized weapons can be carried.

Personally I prefer the latter option - CoP definitely has the best feature set out of all 3 engines, so cutting out some of those features feels counterintuitive to the project. Besides, I've recently returned to playing Stalker, and in my absense I was hoping someone had ported the SoC/CS storylines to the CoP engine so I could enjoy those games with the CoP feature set. Instead we have this, which is equally awesome. Having to actually equip the Psi Helmet to protect us from the Brain Scorcher would be an interesting change to SoC, rather than just having it sit in our inventory.

Tl;dr - Please don't change the way primary/secondary slots work for the SoC/CS builds, same is true for other introduced CoP features. It was challenging enough to replicate some of those via lua scripts for CS, and I look forward to modding CS and eventually SoC to benefit from those features.

No other crashes so far, just some minor tooltip issues - missing text mostly, although the mechanics upgrade tooltip on hover is missing the upgrade price - might be the XML element on the window has been renamed between CS/CoP, so we might be able to fix it with a gamedata tweak. Can also confirm that the upgrade trees are slightly broken. Not too fussed, I feel CoP has the better upgrade trees than CS anyway. CoP gives a bit of build variety, CS is just tree A or tree B.

Regarding Options->Controls, we're missing the keybinding (and ability to edit) for the 'Order Attack' feature, which was introduced in CS, and removed in CoP. Might be an idea to add it in for CS, I was wondering why my K key didn't do anything. Additionally, there are an number of other available keybinds that are missing descriptions in addition to not being editable.

Still working my way through the Swamps, another couple of ingame days before I'm ready to progress into the Cordon.

**Comment by Xottab-DUTY** [comment]
> Regarding Options->Controls, we're missing the keybinding (and ability to edit) for the 'Order Attack' feature, which was introduced in CS

Literally, the game existed for 13 years but only now I have found that there is a command named "Order attack"... 😂
Just returned it.

> Looking at it, I can see two possible resolutions

The solution I see is kinda mix of third solution + both two solutions.
1. The third is to teach the engine to shrink the item texture size to the slot size. This ability is anyway needed, for example, to put something in the quick use slot that has size more than 1x1. (it will crash the original engine with about the same log about insufficient grid space)
2. Add additional logic in the engine to enable/disable CoP-style slots and an option to switch this logic. By default, it will be enabled for CoP and disabled for SoC/CS, but user/modder will be able to override that. It will be a console command or an option in `openxray.ltx`. (or a console command that can be blocked in `openxray.ltx` if eg. modder wants to not give this ability to the user)

> Can also confirm that the upgrade trees are slightly broken.

Yes, the upgrades are not ported fully yet.

> No other crashes so far, just some minor tooltip issues - missing text mostly, although the mechanics upgrade tooltip on hover is missing the upgrade price - might be the XML element on the window has been renamed between CS/CoP, so we might be able to fix it with a gamedata tweak.

That price string is being added by the engine, it's not possible to fix it with gamedata. Will need to fix it :)

Also, can you please tell any other missing text somewhere?

**Comment by globekat** [comment]
> Literally, the game existed for 13 years but only now I have found that there is a command named "Order attack"... 😂
Just returned it.

Its not mentioned in any tutorials, and its kinda buried in the keybinding menu. Dug out my retail copy, and its not even mentioned in the booklet. Unless you get involved in the Faction War aspect and request a squad to attack a location (or tag along with a squad going to attack another location) you won't hear them asking for you to order the attack, and wonder how to order them, and then go looking for how to order them.

> Add additional logic in the engine to enable/disable CoP-style slots and an option to switch this logic. By default, it will be enabled for CoP and disabled for SoC/CS, but user/modder will be able to override that. It will be a console command or an option in openxray.ltx. (or a console command that can be blocked in openxray.ltx if eg. modder wants to not give this ability to the user)

I support a ltx/console flag/setting/command that'll allow the user to switch between CoP and SoC/CS features if there is a difference in implementation between games. Upgrades are one feature that could use a flag. Artifacts are a 2nd (In SoC they just sit around in the open, no detectors needed). And I suppose Clear Sky's infamous Invincible Invisible Bloodsuckers also warrant a mention... [1.5 source code](https://github.com/OpenXRay/xray-15/blob/xd_dev/cs/engine/xrGame/ai/monsters/bloodsucker/bloodsucker.cpp#L677)

> Also, can you please tell any other missing text somewhere?

Trade window, right mouse click on an item in your inventory and the context menu shows 'st_donate'
Enter into dialog with an NPC, place mouse over 'Trade' button, hint window shows 'ui_st_trade_hint'. Also shows when in the trade menu. Mechanics are the same, except the hint window shows 'ui_st_upgrade_hint'.

These are the only 3 instances I've noticed so far ingame. There might be more. I'll keep looking.

**Comment by 1024mb** [comment]
How you guys are launching the game with high resolutions? Mine crashed every time I changed it to 1080p.

**Comment by Xottab-DUTY** [comment]
Just change the resolution from the game console with console command `vid_mode 1920x1080` (that `x` letter is required), then type `cfg_save` and restart the game. :)

**Comment by aodtonix** [comment]
How to change FOV in CS? There's no option in gameplay like in COP.

Edit: Theres a console command 'fov'.

**Comment by Xottab-DUTY** [comment]
New gamedata (fixes #1073 and #1094)

[gamedata.zip](https://github.com/OpenXRay/xray-16/files/12452881/gamedata.zip)

**Comment by blank-teer** [comment]
how does it feel with faction warfare?

**Comment by Xottab-DUTY** [comment]
> how does it feel with faction warfare?

It feels like Clear Sky ;)

**Comment by blank-teer** [comment]
oh man, what a cursed topic

**Comment by donizettilorenzo** [comment]
Any updates for CS and the new engine?

**Comment by Xottab-DUTY** [comment]
Thanks to @St4lker0k765 we now know the reason of weird stripes in PDA:
It's due to incorrectly configured `ui_pda2_fr2_e` texture.

The commit with a gamedata fix: https://github.com/ixray-team/ixray-1.6-stcop/commit/9a706f1eb6de7bd12c92ea5de8a73faeb8708777.

**Comment by Xottab-DUTY** [comment]
> Thanks to [@St4lker0k765](https://github.com/St4lker0k765) we now know the reason of weird stripes in PDA: It's due to incorrectly configured `ui_pda2_fr2_e` texture.
> 
> The commit with a gamedata fix: [ixray-team/ixray-1.6-stcop@9a706f1](https://github.com/ixray-team/ixray-1.6-stcop/commit/9a706f1eb6de7bd12c92ea5de8a73faeb8708777).

It turned out that the texture was configured correctly. `ui_pda2_fr2` texture fragments were laying in different files (texture atlases).
This was supported in CS, but CUIFrameLineWnd was simplified in COP and this feature was removed.
